parsec/scheduling.c

```diff
@@ -186,8 +186,11 @@ int __parsec_execute( parsec_execution_stream_t* es,
                 task->status = PARSEC_TASK_STATUS_COMPLETE;
                 if(PARSEC_DEV_RECURSIVE >= tc->incarnations[chore_id].type) {
                     /* accelerators count their own executed tasks */
+#if 0
+                    /* This is wrong: chore_id is the index in incarnations, but it's not the device id */
                     parsec_device_module_t *dev = parsec_mca_device_get(chore_id);
                     parsec_atomic_fetch_inc_int64((int64_t*)&dev->executed_tasks);
+#endif
```

Comment:
Why is that disabled?

Reply:
Because in `__parsec_execute`  (scheduling.c:138-141) we do
```
   /* Find first bit in chore_mask that is not 0 */
    for(chore_id = 0; NULL != tc->incarnations[chore_id].hook; chore_id++)
        if( 0 != (task->chore_mask & (1<<chore_id)) )
            break;
```

The way I understand this, this finds the first TYPE of incarnation that we want to execute. If I have X CPUs, Y NVIDIA cards and Z Intel cards, incarnations can hold 3 entries, in any order, not X+Y+Z entries.
Then, once we have chosen the type, the evaluate can decide to skip the type, and then the hook can call `get_best_device()` to chose which device between the Y NVIDIA cards that are available.
BUT
Later in the file, at line 189, we do
```
parsec_device_module_t *dev = parsec_mca_device_get(chore_id);
parsec_atomic_fetch_inc_int64((int64_t*)&dev->executed_tasks);
```
I think this is erroneous, and I think  we don't have the device id at this time, it's lost within hook  (which calls `parsec_get_best_device().`

Reply:
This counts only CPU and RECURSIVE, not GPUs (it's `>=`, not `<=`). The GPU accounting is done separately in `device_gpu.c`

Reply:
Right. This should be fixed in PR #616 . If you approve PR #616 and we merge it, I'll rebase and remove that part.

Reply:
see #616 that fixes the underlying problem

Reply:
I merged 616, this will need rebasing